sparse_strips/vello_api/src/pixmap.rs

```diff
@@ -57,106 +88,138 @@ impl Pixmap {
+            pixel.b = premultiply(pixel.b);
+        }
 
-        Ok(Self {
-            width: info
-                .width
-                .try_into()
-                .map_err(|_| png::DecodingError::LimitsExceeded)?,
-            height: info
-                .height
-                .try_into()
-                .map_err(|_| png::DecodingError::LimitsExceeded)?,
-            buf: premultiplied,
-        })
+        Ok(pixmap)
     }
 
     /// Returns a reference to the underlying data as premultiplied RGBA8.
-    pub fn data(&self) -> &[u8] {
+    ///
+    /// The pixels are in row-major order.
+    pub fn data(&self) -> &[PremulRgba8] {
         &self.buf
     }
 
     /// Returns a mutable reference to the underlying data as premultiplied RGBA8.
-    pub fn data_mut(&mut self) -> &mut [u8] {
+    ///
+    /// The pixels are in row-major order.
+    pub fn data_mut(&mut self) -> &mut [PremulRgba8] {
         &mut self.buf
     }
 
+    /// Returns a reference to the underlying data as premultiplied RGBA8.
+    ///
+    /// The pixels are in row-major order. Each pixel consists of four bytes in the order
+    /// `[r, g, b, a]`.
+    pub fn data_as_u8_slice(&self) -> &[u8] {
+        bytemuck::cast_slice(&self.buf)
+    }
+
+    /// Returns a mutable reference to the underlying data as premultiplied RGBA8.
+    ///
+    /// The pixels are in row-major order. Each pixel consists of four bytes in the order
+    /// `[r, g, b, a]`.
+    pub fn data_as_u8_slice_mut(&mut self) -> &mut [u8] {
+        bytemuck::cast_slice_mut(&mut self.buf)
+    }
+
     /// Sample a pixel from the pixmap.
+    ///
+    /// The pixel data is [premultiplied RGBA8][PremulRgba8].
     #[inline(always)]
-    pub fn sample(&self, x: u16, y: u16) -> &[u8] {
-        let idx = 4 * (self.width as usize * y as usize + x as usize);
-        &self.buf[idx..][..4]
+    pub fn sample(&self, x: u16, y: u16) -> PremulRgba8 {
+        let idx = self.width as usize * y as usize + x as usize;
+        self.buf[idx]
+    }
+
+    /// Consume the pixmap, returning the data as the underlying [`Vec`] of premultiplied RGBA8.
+    ///
+    /// The pixels are in row-major order.
+    pub fn take(self) -> Vec<PremulRgba8> {
+        self.buf
     }
 
-    /// Convert from premultiplied to separate alpha.
+    /// Consume the pixmap, returning the data as (unpremultiplied) RGBA8 bytes.
     ///
     /// Not fast, but useful for saving to PNG etc.
-    #[expect(
-        clippy::cast_possible_truncation,
-        reason = "cannot overflow in this case"
-    )]
-    pub fn unpremultiply(&mut self) {
-        for rgba in self.buf.chunks_exact_mut(4) {
-            let alpha = 255.0 / rgba[3] as f32;
-
-            if rgba[3] != 0 {
-                rgba[0] = (rgba[0] as f32 * alpha + 0.5) as u8;
-                rgba[1] = (rgba[1] as f32 * alpha + 0.5) as u8;
-                rgba[2] = (rgba[2] as f32 * alpha + 0.5) as u8;
-            }
-        }
+    ///
```

Comment:
Are there any precision concerns in this operation, incidentally?

Reply:
Yes, premultiplying and unpremultiplying in `u8` is a bit lossy.